c++/src/kj/mutex.c++

```diff
@@ -263,6 +288,10 @@ bool Mutex::lock(Exclusivity exclusivity, Maybe<Duration> timeout,
         }
         state = __atomic_load_n(&futex, __ATOMIC_ACQUIRE);
       }
+
+      // First reader wins in terms of saving the location.
```

Comment:
Seems like last reader wins, since every reader will overwrite? You could check if the increment brought the reader count to 1, in order to decide whether this is the "first" reader.

Reply:
"First reader" here means "first reader to hit this". I originally had this check the count but I was uneasy that there would be a non-controllable amount of time where the lock is held by a reader without the state being updated (e.g. unlock with 5k threads blocked on a reader lock, the first lock may be the last one out & take a while). This approach felt safer in that it shortens this time window to be more bounded since the locks trying to set the state will be the batch that's winning the thundering herd.

Reply:
> "First reader" here means "first reader to hit this".

OK, but again, it looks like _every_ reader will overwrite `lockAcquiredLocation`, not just the first reader, so I don't understand how the first one "wins". In fact they could be writing to `lockAcquiredLocation` at the same time so the value could end up inconsistent (like a filename from one thread and a line number from another). That's all fine if this is just for debugging anyway but I think the comment could be clearer.

Reply:
Unfortunately github doesn't make it easy to view diffs between patchsets. I think the comment was from when I was checking for reader count to 1. I thought I had changed it to an atomic that protected the first one but I guess I had changed my mind part-way through. Sorry for the confusion. I'll update the comment.